api/spec/api_docs/definitions/resource.rb

```diff
@@ -1,78 +1,149 @@
+      end
+
+      def create_request
+        definition = make_request(__callee__, create_attributes || request_attributes)
+        definition = DryTypesToJson.convert(definition)
+        transform_keys(definition)
+      end
+
+      def update_attributes
+        nil
+      end
+
+      def create_attributes
+        nil
+      end
+
+      def request_attributes
+        ::Types::Hash.schema(self::REQUEST_ATTRIBUTES.merge(full_attributes).except(*self::READ_ONLY))
+      end
+
+      def request_relationships
+        Type.object(properties: {})
+      end
+
+      def required_create_attributes
+        const_defined?(:REQUIRED_CREATE_ATTRIBUTES) ? self::REQUIRED_CREATE_ATTRIBUTES : required_attributes
+      end
+
+      def required_update_attributes
+        const_defined?(:REQUIRED_UPDATE_ATTRIBUTES) ? self::REQUIRED_UPDATE_ATTRIBUTES : required_attributes
+      end
+
+      def required_attributes
+        const_defined?(:REQUIRED_ATTRIBUTES) ? self::REQUIRED_ATTRIBUTES : []
+      end
+
+      ######################################
+      ############# RESPONSES ##############
+      ######################################
+
+      def resource_data(attr, relation)
+        ::Types::Hash.schema(
+          id: ::Types::Serializer::ID,
+          type: ::Types::String.meta(example: type),
+          attributes: attr,
+          relationships: relation,
+          meta: ::Types::Serializer::Meta
+        )
+      end
+
+      def resource_response_data
+        return collection_response_data if serializer.partial_only?
+
+        resource_data(
+          response_attributes(full_attributes),
+          response_relationships(full_relationships)
         )
+      end
+
+      def resource_response
+        definition = ::Types::Hash.schema(data: resource_response_data)
+
+        definition = DryTypesToJson.convert(definition)
         debug(__callee__, definition)
         transform_keys(definition)
       end
 
       def collection_response
-        definition = Type.object(
-          properties: {
-            data: Type.array(
-              items: collection_resource_response || resource_response
-            )
-          }
+        definition = ::Types::Hash.schema(
+          data: ::Types::Array.of(collection_response_data)
         )
+
+        definition = DryTypesToJson.convert(definition)
         debug(__callee__, definition)
         transform_keys(definition)
       end
 
-      def update_request
-        definition = make_request(__callee__, update_attributes || request_attributes)
-        transform_keys(definition)
+      def collection_response_data
+        resource_data(
+          response_attributes(attributes),
+          response_relationships(relationships)
+        )
       end
 
-      def create_request
-        definition = make_request(__callee__, create_attributes || request_attributes)
-        transform_keys(definition)
+      def response_relationships(relation)
```

Comment:
This method is still named poorly. The name should say what it does. In this case, it looks like it's taking a relationship (don't call things one thing in one place and another thing in another place) and type casting it somehow. If that's what it's doing, call the method `cast_relationship`.